solutions/cses-1160.mdx

```diff
@@ -67,289 +35,114 @@ direction.
-}
-
-// oops doesn't compile with current template LOL ~Ben
-
-int n,q,t[MX], ind[MX], depth[MX], comp[MX], nex[MX];
-int tmp[MX], par[MX][18];
-vi child[MX], adj[MX];
-bool incyc[MX];
-
-// not in cyc, not in cyc
-// not in cyc, in cyc
-// in cyc, not in cyc
-// in cyc, in cyc
-
-void dfs(int x) {
-    int X = x;
-    while (tmp[x] == 0) {
-        tmp[x] = X;
-        x = t[x];
-    }
-    while (tmp[x] == X) {
-        tmp[x] = X+n; ind[x] = nex[comp[x]]++;
-        incyc[x] = 1;
-        x = t[x];
-    }
-}
-
-void DFS(int cur) {
-    for (int i: child[cur]) if (!incyc[i]) {
-        depth[i] = depth[cur]+1;
-        DFS(i);
-    }
-}
-
-void process() {
-    FOR(i,1,n+1) dfs(i);
-    FOR(i,1,n+1) par[i][0] = t[i];
-    FOR(i,1,18) FOR(j,1,n+1)
-        par[j][i] = par[par[j][i-1]][i-1];
-    FOR(i,1,n+1) if (incyc[i]) DFS(i);
-}
-
-int getPar(int a, int b) {
-    R0F(i,18) if (b&(1<<i)) a = par[a][i];
-    return a;
-}
-
-void gencomp(int x, int i) {
-    if (comp[x]) return;
-    comp[x] = i;
-    for (int y: adj[x]) gencomp(y,i);
-}
-
-int main() {
-    ios_base::sync_with_stdio(0); cin.tie(0);
-    cin >> n >> q;
-    FOR(i,1,n+1) {
-        cin >> t[i];
-        child[t[i]].pb(i);
-        adj[i].pb(t[i]), adj[t[i]].pb(i);
-    }
-    FOR(i,1,n+1) if (comp[i] == 0) gencomp(i,i);
-    process();
-    F0R(i,q) {
-        int a,b; cin >> a >> b;
-        if (comp[a] != comp[b]) {
-            cout << "-1\n";
-            continue;
-        }
-        if (!incyc[a]) {
-            if (incyc[b]) {
-                int x = getPar(a,depth[a]);
-                cout << ((ind[b]-ind[x]+nex[comp[a]])%nex[comp[a]])+depth[a];
-            } else {
-                if (getPar(a,depth[a]-depth[b]) == b) cout << depth[a]-depth[b];
-                else cout << -1;
-            }
-        } else {
-            if (incyc[b]) cout << (ind[b]-ind[a]+nex[comp[a]])%nex[comp[a]];
-            else cout << -1;
-        }
-        cout << "\n";
-    }
+		return a;
+	};
+
+	auto cycle_dis = [](int a, int b) {
+		assert(node[a].cycle && node[b].cycle);
+		assert(node[a].cycle_id == node[b].cycle_id);
+		if (node[a].dis <= node[b].dis) {
+			return node[b].dis - node[a].dis;
+		} else {
+			return cycle_len[node[a].cycle_id] - (node[a].dis - node[b].dis);
+		}
+	};
+
+	for (int q = 0; q < Q; q++) {
+		int a, b;
+		cin >> a >> b;
+
+		if (node[a].cycle && node[b].cycle) {
+			if (node[a].cycle_id == node[b].cycle_id) {
+				cout << cycle_dis(a, b) << '\n';
+			} else {
+				cout << -1 << '\n';
+			}
+		} else if (node[a].cycle && !node[b].cycle) {
+			cout << -1 << '\n';
+		} else if (!node[a].cycle && node[b].cycle) {
+			if (node[a].cycle_id == node[b].cycle_id) {
+				assert(node[lift(a, node[a].dis)].cycle);
+				cout << node[a].dis + cycle_dis(lift(a, node[a].dis), b) << '\n';
+			} else {
+				cout << -1 << '\n';
+			}
+		} else if (!node[a].cycle && !node[b].cycle) {
+			if (node[a].dis >= node[b].dis && lift(a, node[a].dis - node[b].dis) == b) {
```

Comment:
seems a bit longer than 80 characters lol

Reply:
![image](https://user-images.githubusercontent.com/67491820/144756663-9776cd6a-2ced-48b2-b0d1-c581209b67aa.png)
There's no text wrapping in the code. 80 chars is not a hard limit